contracts/03-tamagotchi-nft/src/lib.rs

```diff
@@ -1,19 +1,178 @@
+            msg::reply(TmgEvent::Fed, 0)
+                .expect("Error sending tamagotchi variant 'Fed'");
+        },
+        TmgAction::Play => {
+            update_field(&mut tamagotchi.entertained, FILL_PER_ENTERTAINMENT);
+            msg::reply(TmgEvent::Entertained, 0)
+                .expect("Error sending tamagotchi variant 'Entertained'");  
+        },
+        TmgAction::Sleep => {
+            update_field(&mut tamagotchi.rested, FILL_PER_SLEEP);
+            msg::reply(TmgEvent::Slept, 0)
+                .expect("Error sending tamagotchi variant 'Slept'");  
+        },
+        TmgAction::Transfer(actor_id) => {
+            let source_id = msg::source();
+            let mut owner_transfered = false;
+            if tamagotchi.owner == source_id {
+                tamagotchi.owner = actor_id;
+                owner_transfered = true;
+            }
+            if let Some(approved_account) = tamagotchi.approved_account {
+                if approved_account == source_id {
+                    tamagotchi.owner = actor_id;
+                    owner_transfered = true;
+                }
+            }
+            if owner_transfered {
+                msg::reply(TmgEvent::Transferred(actor_id), 0)
+                    .expect("Error in sending reply");
+            }
+        },
+        TmgAction::Approve(actor_id) => {
+            let source_id = msg::source();
+            if tamagotchi.owner == source_id {
+                tamagotchi.approved_account = Some(actor_id);
+                msg::reply(TmgEvent::Approved(actor_id), 0)
+                    .expect("Error in sending reply");
+            }
+        },
+        TmgAction::RevokeApproval => {
+            let source_id = msg::source();
+            if tamagotchi.owner == source_id {
+                tamagotchi.approved_account = None;
+                msg::reply(TmgEvent::ApprovalRevoked, 0)
+                    .expect("Error in sending reply");
+            }
+        }
+    }
 }
 
 #[no_mangle]
 extern fn state() {
     // TODO: 0️⃣ Copy the `handle` function from the previous lesson and push changes to the master branch
+    msg::reply(state_ref(), 0)
+        .expect("Failed to share state");
+}
+
+
+fn blocks_height() -> u64 {
+    exec::block_height() as u64
+}
+
+fn state_ref() -> &'static Tamagotchi {
+    let state = unsafe { TAMAGOTCHI.as_ref() };
+    debug_assert!(state.is_some(), "State is not initialized");
+    unsafe { state.unwrap_unchecked() }
 }
+
+fn state_mut() -> &'static mut Tamagotchi {
+    let state = unsafe { TAMAGOTCHI.as_mut() };
+    debug_assert!(state.is_some(), "State is not initialized");
+    unsafe { state.unwrap_unchecked() } 
+}
+
+fn update_tamagotchi_fields() {
+    let state = state_mut();
+    let blocks_height = blocks_height();
+    state.fed = updated_field_value(
+        state.fed,
+        state.fed_block,
+        HUNGER_PER_BLOCK,
+        blocks_height
+    );
+    state.fed_block = blocks_height;
```

Comment:
`fed_block`(and mates) is to be updated after the ` TmgAction::Feed`  action only.